### PR TITLE
Add feature flags package

### PR DESCRIPTION
**Pull request by enejb**
Fixes #

## Proposed changes
* Add a new `automattic/jetpack-feature-flags` package for lightweight feature flag registration and checks.
* Add the `Feature_Flags` registry with `register()`, `get()`, `all()`, and `is_enabled()`.
* Resolve feature state through the shared `jetpack_feature_flag_enabled` filter so storage and rollout controls can live outside the package.
* Add README documentation, PHPUnit scaffolding, unit tests, and a changelog entry.

## Related product discussion/links
* None.

## Does this pull request change what data or activity we track or use?
No. This adds a PHP utility package and does not add tracking or data collection.

## Testing instructions
* Review the package API in `projects/packages/feature-flags/src/class-feature-flags.php`.
* Confirm flags can be registered with defaults and metadata via `Feature_Flags::register()`.
* Confirm `Feature_Flags::is_enabled()` applies the `jetpack_feature_flag_enabled` filter with the default, flag name, and definition.
* Syntax checks were run with `php -l` for the package class and test file.
* PHPUnit was not run locally because `jp` is not installed in this shell, and the `pnpm jetpack` fallback attempted restricted network dependency fetching.

### CODE REVIEW

File: projects/packages/feature-flags/tests/php/Feature_Flags_Test.php [hunk 1]
```diff
@@ -0,0 +1,160 @@
+<?php
```
Comment (anomiex) [inline projects/packages/feature-flags/tests/php/Feature_Flags_Test.php:1]: Add a `projects/packages/feature-flags/tests/.phpcs.dir.xml` with the following contents to stop these incorrect warnings.
```xml
<?xml version="1.0"?>
<ruleset>
	<rule ref="Jetpack-Tests" />
</ruleset>
```

(if you did try to follow these, you'd get a correct error from `Jetpack.PHPUnit.TestClassName.DoesNotMatchFileName` telling you to change it back)

File: projects/packages/feature-flags/src/class-feature-flags.php [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+		self::$flags = array();
+	}
+
+	/**
+	 * Normalize a flag name.
+	 *
+	 * @param string $name Flag name.
+	 * @return string Normalized flag name, or empty string for invalid input.
+	 */
+	private static function normalize_name( $name ) {
```
Comment (anomiex) [inline projects/packages/feature-flags/src/class-feature-flags.php]: The logic of this function seems suspect to me. Only one caller checks for the error case at all, and it does so only to silently fail.

Better, IMO, would be to just throw if an invalid name is passed, and not transform the passed name in any way otherwise. The only situation where you might want normalization is when processing a user input, and in that case the normalization would be better done in the input processing layer rather than here.

Comment (enejb) [reply projects/packages/feature-flags/src/class-feature-flags.php]: Good call. we don't need it!

File: projects/packages/feature-flags/src/class-feature-flags.php [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+				'name'        => $name,
+			);
+		}
+
+		$default = (bool) $definition['default'];
+
+		/**
+		 * Filters whether a Jetpack feature flag is enabled.
+		 *
+		 * @since 0.1.0
```
[the remark this answers is not in the document]
Comment (enejb) [reply projects/packages/feature-flags/src/class-feature-flags.php]: Done in e303686d — switched the filter DocBlock to the `$$next-version$$` placeholder so release tooling fills it in.

File: projects/packages/feature-flags/phpunit.11.xml.dist [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+	beStrictAboutOutputDuringTests="true"
+	failOnRisky="true"
+	failOnWarning="true"
+>
+	<testsuites>
+		<testsuite name="main">
+			<directory suffix="Test.php">tests/php</directory>
+		</testsuite>
+	</testsuites>
+</phpunit>
```
[the remark this answers is not in the document]
Comment (enejb) [reply projects/packages/feature-flags/phpunit.11.xml.dist:36]: Done in e303686d — aligned `phpunit.11.xml.dist` with the standard package config: local schema URL, the stricter `displayDetails*`/`failOn*` settings, and the `<source>`/`<coverage>` blocks.

File: projects/packages/feature-flags/tests/php/Feature_Flags_Test.php (hunk 1 above)
Comment (enejb) [reply projects/packages/feature-flags/tests/php/Feature_Flags_Test.php:1]: Thanks — added `tests/.phpcs.dir.xml` with the `Jetpack-Tests` ruleset in e303686d. PHPCS is clean on the tests now.

File: projects/packages/feature-flags/src/class-feature-flags.php (hunk 2 above)
Comment (enejb) [reply projects/packages/feature-flags/src/class-feature-flags.php]: Followed up on this in 37425728: dropped `validate_name()` entirely (no runtime validation or transformation), and instead added a `Jetpack.FeatureFlags.FeatureFlagName` PHPCS sniff to the codesniffer package that flags invalid literal names passed to `Feature_Flags::register()` at lint time. Dynamic names (variables/constants/concatenation) can't be checked statically, so the README and sniff docs note that callers should normalize untrusted input before registering.

File: projects/packages/codesniffer/Jetpack/Sniffs/FeatureFlags/FeatureFlagNameSniff.php [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+		// We only care about the `register` method name.
+		if ( 'register' !== $tokens[ $stackPtr ]['content'] ) {
+			return;
+		}
+
+		// It must be a static call on the Feature_Flags class: `Feature_Flags::register`.
+		$doubleColon = $phpcsFile->findPrevious( Tokens::$emptyTokens, $stackPtr - 1, null, true );
+		if ( false === $doubleColon || T_DOUBLE_COLON !== $tokens[ $doubleColon ]['code'] ) {
+			return;
```
Comment (anomiex) [inline projects/packages/codesniffer/Jetpack/Sniffs/FeatureFlags/FeatureFlagNameSniff.php:73]: No test covers this code path.

File: projects/packages/codesniffer/Jetpack/Sniffs/FeatureFlags/FeatureFlagNameSniff.php [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+			|| T_STRING !== $tokens[ $className ]['code']
+			|| self::CLASS_NAME !== $tokens[ $className ]['content']
+		) {
+			return;
+		}
+
+		// It must actually be a call (followed by an opening parenthesis).
+		$openParen = $phpcsFile->findNext( Tokens::$emptyTokens, $stackPtr + 1, null, true );
+		if ( false === $openParen || T_OPEN_PARENTHESIS !== $tokens[ $openParen ]['code'] ) {
+			return;
```
Comment (anomiex) [inline projects/packages/codesniffer/Jetpack/Sniffs/FeatureFlags/FeatureFlagNameSniff.php:86]: No test covers this code path.

File: projects/packages/codesniffer/Jetpack/Sniffs/FeatureFlags/FeatureFlagNameSniff.php [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+		}
+
+		// Grab the first parameter ($name).
+		$parameters = PassedParameters::getParameters( $phpcsFile, $stackPtr );
+		if ( ! isset( $parameters[1] ) ) {
+			return;
```
Comment (anomiex) [inline projects/packages/codesniffer/Jetpack/Sniffs/FeatureFlags/FeatureFlagNameSniff.php:92]: No test covers this code path.

File: projects/packages/feature-flags/src/class-feature-flags.php [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+	 * Register a feature flag.
+	 *
+	 * Supported definition keys:
+	 * - default: bool Whether the flag is enabled by default.
+	 * - description: string Human-readable description.
+	 * - owner: string Owning package, plugin, or product area.
+	 *
+	 * Flag names must match /^[a-z0-9][a-z0-9_-]*$/. This is enforced at lint
+	 * time by the `Jetpack.FeatureFlags.FeatureFlagName` PHPCS sniff rather than
+	 * at runtime, so registration stays allocation-free on the hot path.
```
Comment (anomiex) [inline projects/packages/feature-flags/src/class-feature-flags.php:34]: Meh. If it's really a restriction, it would probably be worth enforcing. If it's just a "you really should limit it" to the point we can leave it to an external linting rule (that uses in other repos might not enable), I wonder whether it's worth even linting for.

File: projects/packages/feature-flags/phpunit.12.xml.dist
Comment (anomiex) [inline projects/packages/feature-flags/phpunit.12.xml.dist:1]: All other projects symlink the file to phpunit.11.xml.dist, since they're the same. Any reason not to do that here?

Comment (anomiex) [reply projects/packages/feature-flags/phpunit.12.xml.dist:1]: Also, you should need a `phpunit.8.xml.dist` symlink for the PHP 7.2 run. But apparently https://github.com/Automattic/jetpack/pull/49017 broke running package tests on 7.2; https://github.com/Automattic/jetpack/pull/49832 should fix that.

File: projects/packages/feature-flags/src/class-feature-flags.php (hunk 8 above)
Comment (enejb) [reply projects/packages/feature-flags/src/class-feature-flags.php:34]: I would love for our internal repos to name things in a consistent way but if other folks want to name things in other ways I think it should be fine.

**Comment by dhasilva** [comment]
## PR Review: Add feature flags package

### Summary
Adds a new `automattic/jetpack-feature-flags` package: a tiny static registry (`register`/`get`/`all`/`is_enabled`/`reset`) that resolves every flag through a single `jetpack_feature_flag_enabled` filter and stores no state itself. A companion `Jetpack.FeatureFlags.FeatureFlagName` PHPCS sniff validates literal flag names at lint time. Nothing in the codebase consumes the package yet — this is pure infrastructure.

Reviewed at head `2664b013d5bc80bd1bb225c3d313f48b94fc6115`.

### Changelog / PR description
Both look good. One changelog entry per package, both `Significance: minor / Type: added`, properly punctuated. PR description is accurate, testing instructions present, data/privacy answered, no confidential URLs.

### Backward compatibility
No concerns — pure additions, nothing removed, renamed, or re-signatured.

### Findings

**Design / API**

1. `[suggestion]` **No per-flag dynamic filter — overriding loses the `__return_*` one-liner.** Every flag routes through the single `jetpack_feature_flag_enabled` filter, so an override snippet must branch on `$flag_name`; you can't `add_filter( 'jetpack_feature_flag_enabled', '__return_true' )` without flipping *every* flag. Firing an additional dynamic filter would restore that ergonomic and match WP conventions (`get_option` → `option_{$name}`):
   ```php
   $enabled = (bool) apply_filters( 'jetpack_feature_flag_enabled', $default, $name, $definition );
   /** This filter is documented above; dynamic variant keyed by flag name. */
   return (bool) apply_filters( "jetpack_feature_flag_enabled_{$name}", $enabled, $definition );
   ```
   https://github.com/Automattic/jetpack/blob/2664b013d5bc80bd1bb225c3d313f48b94fc6115/projects/packages/feature-flags/src/class-feature-flags.php#L107

2. `[suggestion]` **`all()` mutates the canonical store as a side effect.** `ksort( self::$flags )` re-orders the live static array on every read. A getter sorting global state is surprising; sort a copy instead:
   ```php
   public static function all() {
       $flags = self::$flags;
       ksort( $flags );
       return $flags;
   }
   ```
   https://github.com/Automattic/jetpack/blob/2664b013d5bc80bd1bb225c3d313f48b94fc6115/projects/packages/feature-flags/src/class-feature-flags.php#L83-L88

3. `[suggestion]` **Default-definition array is duplicated.** The unknown-flag branch in `is_enabled()` re-literals the same defaults built in `register()`. Extracting a small `private static function default_definition( $name )` keeps them in sync (and honors the "allocation-free hot path" note — `is_enabled` allocates a fresh array per call for unknown flags).
   https://github.com/Automattic/jetpack/blob/2664b013d5bc80bd1bb225c3d313f48b94fc6115/projects/packages/feature-flags/src/class-feature-flags.php#L100-L107

4. `[suggestion]` **"Mechanism, no policy" means flags are static defaults until someone writes the filter.** The package ships no storage, admin toggle, percentage rollout, or constant/query override, so until a policy layer hooks the filter, `is_enabled()` only ever returns the registered `default`. Worth confirming the consuming/rollout PR is planned so this doesn't land as a dead utility.

**Sniff**

5. `[suggestion]` **Class match is the unqualified literal `Feature_Flags` only.** Aliased imports (`use … as FF; FF::register()`) are silently skipped (false negative), and any unrelated `Feature_Flags::register()` elsewhere would be matched (theoretical false positive). Fine for a lint-time convenience — worth a one-line comment noting the limitation.
   https://github.com/Automattic/jetpack/blob/2664b013d5bc80bd1bb225c3d313f48b94fc6115/projects/packages/codesniffer/Jetpack/Sniffs/FeatureFlags/FeatureFlagNameSniff.php#L60-L66

6. `[suggestion]` **Lint-only validation leaves dynamic names unguarded at runtime.** Names become array keys and filter args, so a malformed dynamic name isn't dangerous, just silently inconsistent (the README's "normalize before registering" is the only safety net). Acceptable, but a cheap `WP_DEBUG`-only `_doing_it_wrong()` guard in `register()` would catch dynamic mistakes the sniff structurally can't. Optional.

**Tests**

7. `[suggestion]` Coverage is solid (register/get, defaults, sorting, default resolution, filter override, unknown-flag + filter). Minor gaps worth a line each: non-bool `default` is cast to bool, `get()` returns `null` for unknown, and re-registering overwrites. All low priority.

### Other checks
Security (no REST/SQL/caps/output), performance, error handling, translations (no user-facing strings), naming, composer conventions (matches sibling `status` package), and sniff auto-discovery: all clear. PHP `>=7.2` is satisfied (`??` only).

### Test results
- PHP: **Not executed.** Runner: `jp test php packages/feature-flags`. `php -l` on `class-feature-flags.php` and the test file reports no syntax errors.
- Phan: Not executed. Config present. Runner: `jp phan packages/feature-flags`.

### Verdict
**Minor issues — can merge after addressing.** No blockers: low-risk, purely additive infrastructure with good tests and correct conventions. The suggestions worth a reply are #1 (dynamic per-flag filter) and #4 (confirming a consumer/policy layer is coming).

_Generated by Claude._

File: projects/packages/codesniffer/Jetpack/Sniffs/FeatureFlags/FeatureFlagNameSniff.php (hunk 5 above)
Comment (enejb) [reply projects/packages/codesniffer/Jetpack/Sniffs/FeatureFlags/FeatureFlagNameSniff.php:73]: Covered in 973956a4 — the sniff fixture now includes a non-static call, `$thing->register( 'Not Static Here!' )`, which lacks a `T_DOUBLE_COLON` and hits this early return.

File: projects/packages/codesniffer/Jetpack/Sniffs/FeatureFlags/FeatureFlagNameSniff.php (hunk 6 above)
Comment (enejb) [reply projects/packages/codesniffer/Jetpack/Sniffs/FeatureFlags/FeatureFlagNameSniff.php:86]: Covered in 973956a4 — `Feature_Flags::register;` (class-constant access, no opening paren) exercises this return.

File: projects/packages/codesniffer/Jetpack/Sniffs/FeatureFlags/FeatureFlagNameSniff.php (hunk 7 above)
Comment (enejb) [reply projects/packages/codesniffer/Jetpack/Sniffs/FeatureFlags/FeatureFlagNameSniff.php:92]: Covered in 973956a4 — `Feature_Flags::register()` with no arguments exercises this return.

File: projects/packages/feature-flags/phpunit.12.xml.dist
Comment (enejb) [reply projects/packages/feature-flags/phpunit.12.xml.dist:1]: Good call — done in 973956a4. `phpunit.12.xml.dist` is now a symlink to `phpunit.11.xml.dist`, and I added `phpunit.8.xml.dist` -> `phpunit.9.xml.dist` for the PHP 7.2 run. Thanks for the pointer to #49017 / #49832.

**Comment by enejb** [comment]
@dhasilva thanks for the thorough pass. Updates in c53c3d45:

1. **Dynamic per-flag filter** — added `jetpack_feature_flag_enabled_{$name}` alongside the global filter, so a single flag can be flipped with a `__return_true`/`__return_false` one-liner. Covered by a new test.
2. **`all()` mutating the store** — now sorts a copy instead of `ksort()`-ing the live static array.
3. **Duplicated defaults** — extracted into a private `default_definition()` used by both `register()` and `is_enabled()`.
4. **Consumer/policy layer** — yes, a follow-up PR will hook the filter into a real feature; landing the mechanism standalone first is intentional so consumers can adopt it incrementally.
5. **Sniff matches the unqualified literal only** — correct, and intentional: aliased imports and dynamic names are skipped (false-negative-safe). It's a lint-time convenience, with the README's "normalize untrusted input before registering" as the runtime contract.
6. **Runtime `_doing_it_wrong()` guard** — leaving this out on purpose to keep `register()` allocation-free on the hot path; validation lives in the sniff per the earlier thread with anomiex.
7. **Extra test cases** (non-bool `default` cast, `get()` → `null`, re-register overwrite) — happy to add these as quick one-liners if you'd like them in before merge.

**Comment by manzoorwanijk** [comment]
> Kill switch — ship a risky feature behind a flag so it can be disabled in production without a deploy.

How?

**Comment by enejb** [comment]
> How?

Good question. I added a bit more context there. This part is still not implemented yet but I can see us adding a feature on simple sites where blog stickers could be used to test a feature that is behind a feature flag. 

I think this would be a good way to do this when the feature is still early in the deploy process.

**Comment by manzoorwanijk** [comment]
> Good question. I added a bit more context there. This part is still not implemented yet but I can see us adding a feature on simple sites where blog stickers could be used to test a feature that is behind a feature flag.
> 
> I think this would be a good way to do this when the feature is still early in the deploy process.

For Social, that is exactly what we have been doing all this while (e.g. #49619) and it works very well and allows us to flip the switch after release when support and other docs etc. are ready.

**Comment by enejb** [comment]
> For Social, that is exactly what we have been doing all this while (e.g. https://github.com/Automattic/jetpack/pull/49619) and it works very well and allows us to flip the switch after release when support and other docs etc. are ready.

Can you explain a bit how this works? 
We want to be able to release this feature slowly to 10%, 50% , 100% of users (on simple and atomic) 

Would we be able to do this using the wpcom site feature function?

**Comment by manzoorwanijk** [comment]
> Can you explain a bit how this works? We want to be able to release this feature slowly to 10%, 50% , 100% of users (on simple and atomic)
> 
> Would we be able to do this using the wpcom site feature function?

If you check `Store_Product_List::get_site_specific_features_data()` on wpcom, you can see how we control the feature based on the plugin versions. See 221018-ghe-Automattic/wpcom as an example.

I am sure there is a way to control it for the percentage of sites.
